### PR TITLE
hw3

### PR DESCRIPTION
**Pull request by cadmiumyellow**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -3,14 +3,147 @@
...
+            <artifactId>selenium-java</artifactId>
+            <version>3.141.59</version>
+        </dependency>
+    </dependencies>
     <properties>
         <maven.compiler.source>8</maven.compiler.source>
         <maven.compiler.target>8</maven.compiler.target>
+        <java.version>1.8</java.version>
+        <encoding>UTF-8</encoding>
+        <test.suites>${project.basedir}\src/test/resources</test.suites>
```
Comment (DmitryKhodakovsky) [inline pom.xml]: `<test.suites>${project.basedir}\src/test/resources</test.suites>` -> `<test.suites>${project.basedir}/src/test/resources</test.suites>`

File: pom.xml [hunk 2]
```diff
@@ -3,14 +3,147 @@
...
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <version>3.0.0-M5</version>
+                <!--<configuration>
+                    <suiteXmlFiles>
+                        <suiteXmlFile>${testXML}</suiteXmlFile>
+                        &lt;!&ndash;<suiteXmlFile>${test.suites}</suiteXmlFile>&ndash;&gt;
+                    </suiteXmlFiles>
+                </configuration>-->
```
Comment (DmitryKhodakovsky) [inline pom.xml]: Why configuration is commented?

File: src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    private WaitActions waitActions;
+    private WebDriver webDriver;
+    private WebElement webElement;
+
+
+    @FindBy(css = ".label-checkbox:first-child > input")
+    private WebElement waterCheckbox;
+
+    @FindBy(css = ".label-checkbox:nth-last-child(2) > input")
+    private WebElement windCheckbox;
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java:18]: Checkboxes could be represented as `List<WebElement>`

File: src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    private WebElement colorsDropDown;
+
+    @FindBy(xpath = "//*[text()='Yellow']")
+    private WebElement yellow;
+
+    @FindBy(css = "ul.panel-body-list > li")
+    private List<WebElement> logs;
+
+
+    @Inject
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java:33]: What if the purpose of the using of current annotation?

File: src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import org.openqa.selenium.support.PageFactory;
+
+import javax.inject.Inject;
+import java.util.List;
+
+public class DifferentElementsPage {
+
+    private WaitActions waitActions;
+    private WebDriver webDriver;
+    private WebElement webElement;
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java]: Why do you need this element?

File: src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.PageFactory;
+
+import javax.inject.Inject;
+import java.util.List;
+
+public class DifferentElementsPage {
+
+    private WaitActions waitActions;
+    private WebDriver webDriver;
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java]: I suppose that these fields could be extracted to the **Base class**

File: src/test/java/ru/training/at/hw3/page/HomePage.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+import org.openqa.selenium.support.PageFactory;
+
+import javax.inject.Inject;
+import java.util.List;
+
+public class HomePage {
+
+    private WaitActions waitActions;
+    private WebDriver webDriver;
+    private WebElement webElement;
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw3/page/HomePage.java]: Code duplication from **src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java**

File: src/test/java/ru/training/at/hw3/page/HomePage.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    private List<WebElement> leftMenuItems;
+
+    @FindBy(css = "a.dropdown-toggle > span.caret:first-child")
+    private WebElement serviceDropDown;
+
+    @FindBy(css = "header ul.dropdown-menu a[href='different-elements.html']")
+    private WebElement diffElementsPage;
+
+
+    @Inject
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw3/page/HomePage.java:61]: Why do you need this annotation?

File: src/test/java/ru/training/at/hw3/page/HomePage.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    public void login() {
+        loginDropdown.click();
+        userName.sendKeys(new Prop().getPropertyValue("userName"));
+        password.sendKeys(new Prop().getPropertyValue("password"));
+        loginButton.click();
+    }
+
+    public Boolean logoutEnabled() {
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw3/page/HomePage.java]: `isLogoutEnabled`

File: src/test/java/ru/training/at/hw3/page/HomePage.java [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        return logoutButton.isEnabled();
+    }
+
+    public Boolean loggedNameDisplayed() {
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw3/page/HomePage.java]: `isLoggedNameDisplayed`

File: src/test/java/ru/training/at/hw3/page/HomePage.java [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+
+    public List<WebElement> getIcons() {
+        return icons;
+    }
+
+    public List<WebElement> getIconsTexts() {
+        return iconsTexts;
+    }
+
+    public Boolean iframeIsDisplayed() {
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw3/page/HomePage.java]: `isIFrameDisplayed`

File: src/test/java/ru/training/at/hw3/page/HomePage.java [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+        return iframe.isDisplayed();
+    }
+
+    public void switchToIframe() {
+        String windowHandler = webDriver.getWindowHandle();
+        webDriver.switchTo().frame(iframe);
+        //return windowHandler;
+    }
+
+    public Boolean iframeButtonIsDisplayed() {
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw3/page/HomePage.java]: `isIFrameButtonDisplayed`

File: src/test/java/ru/training/at/hw3/page/HomePage.java [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+        return iframeButton.isDisplayed();
+    }
+
+    public void switchBackToHomePage() {
+        webDriver.switchTo().parentFrame();
+    }
+
+    public Boolean leftSectionIsDisplayed() {
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw3/page/HomePage.java]: `isLeftSectionDisplayed`

File: src/test/java/ru/training/at/hw3/test/BaseTest.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+        webDriver.manage().deleteAllCookies();
+    }
+
+    @AfterClass
+    public void resetBetweenTests() {
+        webDriver.manage().deleteAllCookies();
+        webDriver.navigate().refresh();
+    }
+
+    @AfterSuite(alwaysRun = true)
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw3/test/BaseTest.java:59]: I suppose that here should be used `@AfterMethod`. Driver should be closed after each test

**Review by DmitryKhodakovsky — CHANGES_REQUESTED** [review]
Please resolve comments

**Comment by DmitryKhodakovsky** [comment]
Please fix the tests

File: pom.xml (hunk 1 above)
Comment (cadmiumyellow) [reply pom.xml]: done

File: pom.xml (hunk 2 above)
Comment (cadmiumyellow) [reply pom.xml]: I commented it before we talked, now uncommented

File: src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java (hunk 3 above)
Comment (cadmiumyellow) [reply src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java:18]: I just don't need them all, so I chose to implement them individually according to KISS principle

File: src/test/java/ru/training/at/hw3/page/HomePage.java (hunk 9 above)
Comment (cadmiumyellow) [reply src/test/java/ru/training/at/hw3/page/HomePage.java]: fixed

File: src/test/java/ru/training/at/hw3/page/HomePage.java (hunk 10 above)
Comment (cadmiumyellow) [reply src/test/java/ru/training/at/hw3/page/HomePage.java]: fixed

File: src/test/java/ru/training/at/hw3/page/HomePage.java (hunk 11 above)
Comment (cadmiumyellow) [reply src/test/java/ru/training/at/hw3/page/HomePage.java]: due to checkstyle corrected it to `isIframeDisplayed `

File: src/test/java/ru/training/at/hw3/page/HomePage.java (hunk 12 above)
Comment (cadmiumyellow) [reply src/test/java/ru/training/at/hw3/page/HomePage.java]: fixed

File: src/test/java/ru/training/at/hw3/page/HomePage.java (hunk 13 above)
Comment (cadmiumyellow) [reply src/test/java/ru/training/at/hw3/page/HomePage.java]: also fixed

Comment (cadmiumyellow) [reply src/test/java/ru/training/at/hw3/page/HomePage.java]: fixed

File: src/test/java/ru/training/at/hw3/page/HomePage.java (hunk 10 above)
Comment (cadmiumyellow) [reply src/test/java/ru/training/at/hw3/page/HomePage.java]: fixed

File: src/test/java/ru/training/at/hw3/page/HomePage.java (hunk 12 above)
Comment (cadmiumyellow) [reply src/test/java/ru/training/at/hw3/page/HomePage.java]: fixed

File: src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java (hunk 4 above)
Comment (cadmiumyellow) [reply src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java:33]: If I don't use this annotation I have a com.google.inject.ConfigurationException: Guice configuration errors and as a result only the FirstTest runs well (the Home page one), and the 2nd (related to this page - DifferentElementsPage) fails.
I am not sure I understand this magic completely, but I believe this is how I have the driver here on this page.

File: src/test/java/ru/training/at/hw3/page/HomePage.java (hunk 11 above)
Comment (cadmiumyellow) [reply src/test/java/ru/training/at/hw3/page/HomePage.java]: fixed

File: src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java (hunk 5 above)
Comment (cadmiumyellow) [reply src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java]: you are right, I don't need it. Deleted

File: src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java (hunk 6 above)
Comment (cadmiumyellow) [reply src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java]: I used to have a BasePage class but I thought it was too small to keep it. Now restored

File: src/test/java/ru/training/at/hw3/page/HomePage.java (hunk 7 above)
Comment (cadmiumyellow) [reply src/test/java/ru/training/at/hw3/page/HomePage.java]: extracted to BasePage, no duplication any more

File: src/test/java/ru/training/at/hw3/page/HomePage.java (hunk 8 above)
Comment (cadmiumyellow) [reply src/test/java/ru/training/at/hw3/page/HomePage.java:61]: just like that one on DifferentElementsPage

File: src/test/java/ru/training/at/hw3/test/BaseTest.java (hunk 14 above)
Comment (cadmiumyellow) [reply src/test/java/ru/training/at/hw3/test/BaseTest.java:59]: Unfortunately, it fails with `@AfterMethod` and passes with `@AfterSuite`

File: src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java (hunk 3 above)
Comment (DmitryKhodakovsky) [reply src/test/java/ru/training/at/hw3/page/DifferentElementsPage.java:18]: I am not sure about KISS principal here. Mostly it looks like as YAGNI :)

**Review by DmitryKhodakovsky — APPROVED** [review]
Approved

**Comment by DmitryKhodakovsky** [comment]
Please resolve conflicts and merge it to the master branch. HW3 is aproved

**Pull request merged**